flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SubQueryInputTypeStrategy.java

```diff
@@ -0,0 +1,115 @@
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.FunctionDefinition;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.ArgumentCount;
+import org.apache.flink.table.types.inference.CallContext;
+import org.apache.flink.table.types.inference.ConstantArgumentCount;
+import org.apache.flink.table.types.inference.InputTypeStrategy;
+import org.apache.flink.table.types.inference.Signature;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.table.types.logical.StructuredType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeChecks;
+import org.apache.flink.table.types.logical.utils.LogicalTypeMerging;
+import org.apache.flink.table.types.utils.TypeConversions;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/** {@link InputTypeStrategy} for {@link BuiltInFunctionDefinitions#IN}. */
+@Internal
+public class SubQueryInputTypeStrategy implements InputTypeStrategy {
+    @Override
+    public ArgumentCount getArgumentCount() {
+        return ConstantArgumentCount.from(2);
+    }
+
+    @Override
+    public Optional<List<DataType>> inferInputTypes(
+            CallContext callContext, boolean throwOnFailure) {
+        final LogicalType rightType;
+        final DataType leftType = callContext.getArgumentDataTypes().get(0);
+        if (callContext.getArgumentDataTypes().size() > 2) {
+            final Optional<LogicalType> commonType =
+                    LogicalTypeMerging.findCommonType(
+                            callContext.getArgumentDataTypes().stream()
+                                    .map(DataType::getLogicalType)
+                                    .collect(Collectors.toList()));
+            if (!commonType.isPresent()) {
+                return callContext.fail(
+                        throwOnFailure, "Could not find a common type of the sublist.");
+            }
+            rightType = commonType.get();
+        } else {
+            rightType = callContext.getArgumentDataTypes().get(1).getLogicalType();
+        }
+
+        if (!LogicalTypeChecks.areComparable(
+                leftType.getLogicalType(), rightType, StructuredType.StructuredComparison.EQUALS)) {
+
+            if (!isComparableWithSubQuery(leftType.getLogicalType(), rightType)) {
+                return callContext.fail(
+                        throwOnFailure,
+                        "Types on the right side of IN operator (%s) are not comparable with %s.",
+                        rightType,
+                        leftType.getLogicalType());
+            }
```

Comment:
Are there situations where 
`LogicalTypeChecks.areComparable(left, right)` returns false _and_
`isComparableWithSubQuery(left, right)` returns true?

If not, then it may be more clear to combine the conditions into one `if` statement?

Reply:
Yes, there are. We can still combine that into a single `if` statement.

I added an explanation for the condition.